diff --git a/Makefile b/Makefile
--- a/Makefile
+++ b/Makefile
@@ -23 +23 @@
-src_dir=$(CURDIR)
+src_dir:=$(CURDIR)
@@ -25,3 +25,3 @@
- build_dir=$(shell $(READLINK) -f $(O))
-else
- build_dir=$(CURDIR)/build
+ build_dir:=$(shell $(READLINK) -f $(O))
+else
+ build_dir:=$(CURDIR)/build
@@ -34,3 +34,3 @@
- install_root_dir=$(shell $(READLINK) -f $(I))
-else
- install_root_dir=$(install_root_dir_default)/usr
+ install_root_dir:=$(shell $(READLINK) -f $(I))
+else
+ install_root_dir:=$(install_root_dir_default)/usr
@@ -45 +45 @@
-  platform_dir_path=$(shell $(READLINK) -f $(PLATFORM_DIR))
+  platform_dir_path:=$(shell $(READLINK) -f $(PLATFORM_DIR))
@@ -47 +47 @@
-    platform_parent_dir=$(platform_dir_path)
+    platform_parent_dir:=$(platform_dir_path)
@@ -49,2 +49,2 @@
-    PLATFORM=$(shell basename $(platform_dir_path))
-    platform_parent_dir=$(shell realpath ${platform_dir_path}/..)
+    PLATFORM:=$(shell basename $(platform_dir_path))
+    platform_parent_dir:=$(shell realpath ${platform_dir_path}/..)
@@ -53 +53 @@
- platform_parent_dir=$(src_dir)/platform
+ platform_parent_dir:=$(src_dir)/platform
@@ -68,7 +68,7 @@
-export platform_subdir=$(PLATFORM)
-export platform_src_dir=$(platform_parent_dir)/$(platform_subdir)
-export platform_build_dir=$(build_dir)/platform/$(platform_subdir)
-export include_dir=$(CURDIR)/include
-export libsbi_dir=$(CURDIR)/lib/sbi
-export libsbiutils_dir=$(CURDIR)/lib/utils
-export firmware_dir=$(CURDIR)/firmware
+export platform_subdir:=$(PLATFORM)
+export platform_src_dir:=$(platform_parent_dir)/$(platform_subdir)
+export platform_build_dir:=$(build_dir)/platform/$(platform_subdir)
+export include_dir:=$(CURDIR)/include
+export libsbi_dir:=$(CURDIR)/lib/sbi
+export libsbiutils_dir:=$(CURDIR)/lib/utils
+export firmware_dir:=$(CURDIR)/firmware
@@ -79,6 +79,6 @@
-export KCONFIG_DIR=$(platform_build_dir)/kconfig
-export KCONFIG_AUTOLIST=$(KCONFIG_DIR)/auto.list
-export KCONFIG_AUTOHEADER=$(KCONFIG_DIR)/autoconf.h
-export KCONFIG_AUTOCONFIG=$(KCONFIG_DIR)/auto.conf
-export KCONFIG_AUTOCMD=$(KCONFIG_DIR)/auto.conf.cmd
-export KCONFIG_CONFIG=$(KCONFIG_DIR)/.config
+export KCONFIG_DIR:=$(platform_build_dir)/kconfig
+export KCONFIG_AUTOLIST:=$(KCONFIG_DIR)/auto.list
+export KCONFIG_AUTOHEADER:=$(KCONFIG_DIR)/autoconf.h
+export KCONFIG_AUTOCONFIG:=$(KCONFIG_DIR)/auto.conf
+export KCONFIG_AUTOCMD:=$(KCONFIG_DIR)/auto.conf.cmd
+export KCONFIG_CONFIG:=$(KCONFIG_DIR)/.config
